core/src/main/java/com/linecorp/armeria/common/HttpHeaders.java

```diff
@@ -13,227 +13,157 @@
-     * This method is a shortcut of the following code:
-     * <pre>{@code
-     * headers.names().forEach(name -> {
-     *      if (!contains(name)) {
-     *          set(name, headers.getAll(name));
-     *      }
-     * });
-     * }</pre>
-     */
-    default HttpHeaders setAllIfAbsent(Headers<AsciiString, String, ?> headers) {
-        requireNonNull(headers, "headers");
-        if (!headers.isEmpty()) {
-            headers.names().forEach(name -> {
-                if (!contains(name)) {
-                    set(name, headers.getAll(name));
-                }
-            });
-        }
-        return this;
+    static HttpHeaders of(CharSequence name1, CharSequence value1,
+                          CharSequence name2, CharSequence value2,
+                          CharSequence name3, CharSequence value3,
+                          CharSequence name4, CharSequence value4) {
+        return new HttpHeadersBuilder().add(name1, value1)
+                                       .add(name2, value2)
+                                       .add(name3, value3)
+                                       .add(name4, value4)
+                                       .build();
     }
 
     /**
-     * Returns the immutable view of this headers.
+     * Returns a new {@link HttpHeaders} with the specified headers. The values are converted into
+     * {@link String}s via {@link HttpHeadersBase#addObject(CharSequence, Object)}.
      */
-    default HttpHeaders asImmutable() {
-        return new ImmutableHttpHeaders(this);
+    static HttpHeaders of(CharSequence name1, Object value1,
+                          CharSequence name2, Object value2,
+                          CharSequence name3, Object value3,
+                          CharSequence name4, Object value4) {
+        return new HttpHeadersBuilder().addObject(name1, value1)
+                                       .addObject(name2, value2)
+                                       .addObject(name3, value3)
+                                       .addObject(name4, value4)
+                                       .build();
     }
 
     /**
-     * Returns whether this is immutable or not.
+     * Returns a new builder created from the entries of this headers.
+     *
+     * @see #mutate(Consumer)
      */
-    default boolean isImmutable() {
-        return this instanceof ImmutableHttpHeaders;
-    }
+    HttpHeadersBuilder toBuilder();
 
     /**
-     * Returns a mutable copy of this headers.
-     * If it is already mutable, it returns {@code this}.
+     * Returns a new headers which is the result from the mutation by the specified {@link Consumer}.
+     * This method is a shortcut of:
+     * <pre>{@code
+     * builder = toBuilder();
+     * mutator.accept(builder);
+     * return builder.build();
+     * }</pre>
+     *
+     * @see #toBuilder()
      */
-    default HttpHeaders toMutable() {
-        return this;
+    default HttpHeaders mutate(Consumer<HttpHeadersBuilder> mutator) {
```

Comment:
Possibly `withMutations`?

Anyways just for reference, I think immutablejs provides a good immutable API

https://immutable-js.github.io/immutable-js/docs/#/Map/withMutations

Reply:
Sounds good to me. Thanks for the reference.